File: backend/policy-service/src/main/java/com/insurance/policyservice/service/ClaimService.java
package com.insurance.policyservice.service;

import com.insurance.policyservice.model.Claim;
import com.insurance.policyservice.model.ClaimStatus;
import com.insurance.policyservice.model.Policy;
import com.insurance.policyservice.repository.ClaimRepository;
import com.insurance.policyservice.repository.PolicyRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

@Service
public class ClaimService {
    @Autowired
    private ClaimRepository claimRepository;

    @Autowired
    private PolicyRepository policyRepository;

    public List<Claim> getAllClaims() {
        return claimRepository.findAll();
    }

    public Optional<Claim> getClaimById(Long id) {
        return claimRepository.findById(id);
    }

    public Claim createClaim(Claim claim) {
        // Verify that the policy exists
        Policy policy = policyRepository.findById(claim.getPolicyId())
                .orElseThrow(() -> new RuntimeException("Policy not found with id: " + claim.getPolicyId()));

        // Set initial status
        claim.setStatus(ClaimStatus.PENDING);

        return claimRepository.save(claim);
    }

    public Claim updateClaim(Long id, Claim claimDetails) {
        Claim claim = claimRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Claim not found with id: " + id));

        claim.setDate(claimDetails.getDate());
        claim.setDescription(claimDetails.getDescription());
        claim.setClaimedAmount(claimDetails.getClaimedAmount());
        claim.setSettledAmount(claimDetails.getSettledAmount());
        claim.setStatus(claimDetails.getStatus());

        return claimRepository.save(claim);
    }

    public Claim processClaimStatus(Long id, ClaimStatus status, Double settledAmount) {
        Claim claim = claimRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Claim not found with id: " + id));

        claim.setStatus(status);
        if (settledAmount != null) {
            claim.setSettledAmount(BigDecimal.valueOf(settledAmount)); 
        }

        return claimRepository.save(claim);
    }

    public void deleteClaim(Long id) {
        Claim claim = claimRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Claim not found with id: " + id));

        claimRepository.delete(claim);
    }

    public List<Claim> getClaimsByPolicyId(Long policyId) {
        return claimRepository.findByPolicyId(policyId);
    }
}

File: frontend/src/services/claimService.ts
import api from "./api"
import type { Claim, ClaimStatus } from "../types/Claim"

const CLAIMS_URL = "/claims"

const getAllClaims = async (): Promise<Claim[]> => {
  const response = await api.get<Claim[]>(CLAIMS_URL)
  return response.data
}

const getClaimById = async (id: number): Promise<Claim> => {
  const response = await api.get<Claim>(`${CLAIMS_URL}/${id}`)
  return response.data
}

const getClaimsByPolicyId = async (policyId: number): Promise<Claim[]> => {
  const response = await api.get<Claim[]>(`${CLAIMS_URL}/policy/${policyId}`)
  return response.data
}

const createClaim = async (claim: Claim): Promise<Claim> => {
  const response = await api.post<Claim>(CLAIMS_URL, claim)
  return response.data
}

const updateClaim = async (id: number, claim: Claim): Promise<Claim> => {
  const response = await api.put<Claim>(`${CLAIMS_URL}/${id}`, claim)
  return response.data
}

// const updateClaimStatus = async (id: number, status: ClaimStatus, settledAmount?: number): Promise<Claim> => {
//   const url = settledAmount
//     ? `${CLAIMS_URL}/${id}/status?status=${status}&settledAmount=${settledAmount}`
//     : `${CLAIMS_URL}/${id}/status?status=${status}`

//   const response = await api.patch<Claim>(url)
//   return response.data
// }

const updateClaimStatus = async (id: number, status: ClaimStatus, settledAmount?: number): Promise<Claim> => {
  const response = await api.patch<Claim>(`${CLAIMS_URL}/${id}/status`, { status, settledAmount });
  return response.data;
};


const deleteClaim = async (id: number): Promise<void> => {
  await api.delete(`${CLAIMS_URL}/${id}`)
}

const claimService = {
  getAllClaims,
  getClaimById,
  getClaimsByPolicyId,
  createClaim,
  updateClaim,
  updateClaimStatus,
  deleteClaim,
}

export default claimService

